src/osd/osd_types.cc

```diff
@@ -1084,17 +1084,17 @@ SnapContext pg_pool_t::get_snap_context() const
   return SnapContext(get_snap_seq(), s);
 }
 
-static string make_hash_str(const string &inkey, const string &nspace)
-{
-  if (nspace.empty())
-    return inkey;
-  return nspace + '\037' + inkey;
-}
-
 uint32_t pg_pool_t::hash_key(const string& key, const string& ns) const
 {
-  string n = make_hash_str(key, ns);
-  return ceph_str_hash(object_hash, n.c_str(), n.length());
+ if (ns.empty()) 
+    return ceph_str_hash(object_hash, key.c_str(), key.length());
+  int nsl = ns.length();
+  int len = key.length() + nsl + 1;
+  char buf[len];
+  memcpy(&buf[0], ns.c_str(), nsl);
+  buf[nsl] = '\037';
+  memcpy(&buf[nsl+1], key.c_str(), key.length());
```

Comment:
I think using data() instead of c_str() is better, as there is some possibility that std::string will have to reallocate a copy that has a null terminator for c_str() (but for data() it never will).